upstream/upstream.go

```diff
@@ -655,10 +667,101 @@ func (u *Upstream) shouldSkip(req *common.NormalizedRequest) (reason error, skip
 	}
 
 	// TODO evm: if block can be determined from request and upstream is only full-node and block is historical skip
+	jrReq, err := req.JsonRpcRequest()
+	if err != nil {
+		return fmt.Errorf("failed to get json-rpc request: %w", err), false
+	}
+
+	if index, exists := common.MethodsWithBlockParam[jrReq.Method]; exists {
+		if len(jrReq.Params) > index {
+			// TODO: do we need to check if blockNumber is a valid hex?
+			_, ok := jrReq.Params[index].(string)
+
+			if ok {
+				nodeType := u.config.Evm.NodeType
+				if nodeType == "" {
+					nodeType, err = u.detectNodeType(req)
+					if err != nil {
+						return fmt.Errorf("failed to detect node type: %w", err), false
+					}
+				}
+
+				if nodeType == common.EvmNodeTypeFull {
+					lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
+
+					bn, ebn := req.EvmBlockNumber()
+					if ebn != nil {
+						return fmt.Errorf("failed to get request block number: %w", ebn), false
+					}
+
+					if bn < lb-int64(u.config.Evm.MaxAvailableRecentBlocks) {
+						return fmt.Errorf("block number %d is out of range (must be >= %d)", bn, lb-int64(u.config.Evm.MaxAvailableRecentBlocks)), true
+					}
+				}
+			}
+		}
+	}
```

Comment:
Looking at body of req.EvmBlockNumber() do we even need to manually check with MethodsWithBlockParam?

This method seems to be already taking care of all edge-cases and blockNumber extraction. We can just check if "ebn" is above 0 and then try to fetch the "latestBlock" and so on...